### PR TITLE
Several improvements on PPG + Transformer + MetaDrive

### PR DESCRIPTION
**Pull request by breakds**
## Motivation

There are a few changes on the vectorized MetaDrive experiment that are not committed yet. This PR summarizes them.

## Changes

1. Add `history_frame_skip` for other agents' trajectory inputs. Instead of putting all the frames within the history window into the trajectory, it picks only every `history_frame_skip` frames so that the input is significantly smaller, without much loss of information.
2. Observation includes `map_mask` and `agent_mask`, so that the transformer encoder knows what slots need to be ignored.
3. In `layers.TransformerBlock`, correct the meaning of `mask` by reversing it.

## Testing

Ran experiments to make sure the changes aligns with the expectation (i.e. resembles the previous best experiment result). In the graph below:

blue: `num_evs' = 36
orange: `num_envs` = 24

metadrive_ppg

### CODE REVIEW

File: alf/layers.py [hunk 1]
```diff
@@ -2581,7 +2581,7 @@ def calc_output_shape(self, input_shape):
 
 def _masked_softmax(logits, mask, dim=-1):
     if mask is not None:
-        logits.masked_fill_(mask, -float('inf'))
+        logits.masked_fill_(~mask, -float('inf'))
```
Comment (emailweixu) [inline alf/layers.py]: The doc-string for mask in TransformerBlock says mask is "A tensor for indicating which slot in ``memory`` will NOT be used"

File: alf/environments/metadrive/agent_perception.py [hunk 2]
```diff
@@ -97,6 +97,15 @@ def __init__(self,
         self._history_position = None
         self._history_heading = None
 
+        # Handle Frame Skip
+        H = self._history_window_size
+        self._sampled_index = np.arange(H)[(
```
Comment (emailweixu) [inline alf/environments/metadrive/agent_perception.py]: `np.arange((H-1)%skip, H, skip)`

File: alf/environments/metadrive/agent_perception.py [hunk 3]
```diff
@@ -190,15 +203,18 @@ def observe(self) -> np.ndarray:
         self._visible[:, -1] = self._fov.within(
             transformed_position.point[:, -1])
         self._visible[~alive, -1] = False
+        sampled_visible = self._visible[:, self._sampled_index]
+        sampled_position = transformed_position.point[:, self._sampled_index]
+        sampled_heading = transformed_heading[:, self._sampled_index]
 
         # Shape is [B,]. Denote whether a car is picked to show in the final
         # feature tensor or not. The criterion is that the car has be visible in
```
Comment (emailweixu) [inline alf/environments/metadrive/agent_perception.py]: has be => has been

**Review by emailweixu — CHANGES_REQUESTED** [review]

File: alf/environments/metadrive/agent_perception.py [hunk 4]
```diff
@@ -153,15 +162,19 @@ def reset(self, engine: BaseEngine, ego: BaseVehicle):
         self._history_heading = np.zeros(
             (self._num_agents, self._history_window_size), dtype=np.float32)
 
-    def observe(self) -> np.ndarray:
+    def observe(self) -> Tuple[np.ndarray, int]:
         """Called upon every observation to produce the feature vectors describing the
         dynamic agents that are visible to the ego car. The vectors are
         transformed so that they are in eog car's body frame.
```
Comment (Haichao-Zhang) [inline alf/environments/metadrive/agent_perception.py]: eog car -> ego car

File: alf/layers.py (hunk 1 above)
Comment (breakds) [reply alf/layers.py]: Sorry that I forgot I have already updated the doc-string. Reverted this change and made changes in the configuration instead.

File: alf/examples/metadrive/ppg_metadrive_transformer_conf.py [hunk 5]
```diff
@@ -61,57 +80,85 @@ class ObservationCombiner(torch.nn.Module):
...
+                    ego_feature=EmbeddingConfig(
+                        d_input=input_tensor_spec['ego'].shape[-1],
+                        hidden=(32, )),
+                    agent_feature=EmbeddingConfig(
+                        d_input=input_tensor_spec['agents'].shape[1] *
+                        input_tensor_spec['agents'].shape[2],
+                        hidden=(32, )),
+                    d_model=d_model),
+                input_tensor_spec=input_tensor_spec), lambda x:
+            (x['map_mask'], x['agent_mask'])), lambda x: (x[0], x[1][0], x[1][
```
Comment (emailweixu) [inline alf/examples/metadrive/ppg_metadrive_transformer_conf.py]: This auto-formatting makes it quite hard to read. 
The lambda at line 151 should be indented at the same level ax ObservationCombiner.
The lambda at line 152 should be at a new line and be indented at the same level as MaskedTransformer.
Perhaps use the following to disable the auto-formatting?
```
# yapf: diasble
...
# yapf: enable
```

File: alf/examples/metadrive/ppg_metadrive_transformer_conf.py [hunk 6]
```diff
@@ -61,57 +80,85 @@ class ObservationCombiner(torch.nn.Module):
...
+                    d_model=d_model,
+                    num_heads=num_heads,
+                    memory_size=self._memory_size,
+                    positional_encoding='none'))
+        self._tf_layers = torch.nn.ModuleList(tf_layers)
+
+    def forward(self, inputs):
+        x, map_mask, agent_mask = inputs
+        B = x.shape[0]
+        mask = torch.hstack((torch.ones(B, 1, dtype=bool), map_mask,
```
Comment (emailweixu) [inline alf/examples/metadrive/ppg_metadrive_transformer_conf.py]: torch.ones => torch.zeros

**Review by emailweixu — CHANGES_REQUESTED** [review]

File: alf/examples/metadrive/ppg_metadrive_transformer_conf.py (hunk 5 above)
Comment (breakds) [reply alf/examples/metadrive/ppg_metadrive_transformer_conf.py]: Right, thanks for the advice. Updated.

File: alf/examples/metadrive/ppg_metadrive_transformer_conf.py (hunk 6 above)
Comment (breakds) [reply alf/examples/metadrive/ppg_metadrive_transformer_conf.py]: Thanks for pointing out this bad bug. Fixed.
 I'll run the training again to verify before submitting the PR.

**Comment by breakds** [comment]
Verified that the updated PR still works in training (`num_envs` = 24), with matching performance.

metadrive_verify

**Review by Haichao-Zhang — APPROVED** [review]

**Review by emailweixu — APPROVED** [review]
